<br><br><br>
### Module Title: Applied Data Science with Python 
### Module Code: CSMAD21 
### Lecturer responsible: Dr Carmen Lam and Dr Todd Jones 
### Student Number: 30825102
### Date : 
### Actual hrs spent on the assignment : (expected 40)
<br><br>
<hr style="border:2px solid black">

### Task 1 – Bicycle Journeys: Exploratory Data Analysis

##### The libraries that are essential for the task are imported in the notebook for further use.

In [2]:
import pandas as pd 
import numpy as np

#### As the dataset is of csv format, the read_csv() function from pandas library is used import the dataset. 


In [3]:
metro = pd.read_csv("metro.csv", sep = ',', dtype = {'bike_id': str})

In [5]:
type(metro)

pandas.core.frame.DataFrame

## Encountered Warning:
C:\Users\shrut\AppData\Local\Temp\ipykernel_3676\3546838290.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
Original code: metro = pd.read_csv("metro.csv")
#### Solved by specifying dtype while importing the file.

In [3]:
metro.head()

trip_id  duration     start_time       end_time  start_station  \
0  124657107         5  7/1/2019 0:04  7/1/2019 0:09           4312   
1  124657587         9  7/1/2019 0:07  7/1/2019 0:16           3066   
2  124658068         5  7/1/2019 0:20  7/1/2019 0:25           4410   
3  124659747        20  7/1/2019 0:44  7/1/2019 1:04           3045   
4  124660227        27  7/1/2019 0:44  7/1/2019 1:11           3035   

   start_lat   start_lon  end_station    end_lat     end_lon bike_id  \
0  34.066990 -118.290878         4410  34.063351 -118.296799    6168   
1  34.063389 -118.236160         3066  34.063389 -118.236160   17584   
2  34.063351 -118.296799         4312  34.066990 -118.290878   18920   
3  34.028511 -118.256668         4275  34.012520 -118.285896    6016   
4  34.048401 -118.260948         3049  34.056969 -118.253593    5867   

   plan_duration trip_route_category passholder_type bike_type  
0             30             One Way    Monthly Pass  standard  
1             30          Round Trip    Monthly Pass  electric  
2             30             One Way    Monthly Pass  electric  
3              1             One Way         Walk-up  standard  
4             30             One Way    Monthly Pass  standard

In [4]:
type(metro.iloc[1,10])

str

Why Bike sharing system?
While there are several advantages to this system, it comes with several difficulties for the operators like the optimum demand prediction due to different factors, which leads to a demand and supply problem for the operator, where to build new stations or expand old ones. They also need to know how they can maintain an adequate inventory of bikes.

Natural factors include seasons, months, day of week ,peak timings, working and non-working days, temperature, humidity etc. and man made factors consists of location of bike station, characteristics of the area.

Things I feel may affect the rush:
1. start time & end time (look for peek hrs)
2. Plan duration (pass type)
3. bike type

Things that can be done: (Pre processing)
1. Check rush months
2. Check peek hrs
3. Weekdays vs Weekends
4. Variation depending on season (monsoon more bus)

In [5]:
metro.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92124 entries, 0 to 92123
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   trip_id              92124 non-null  int64  
 1   duration             92124 non-null  int64  
 2   start_time           92124 non-null  object 
 3   end_time             92124 non-null  object 
 4   start_station        92124 non-null  int64  
 5   start_lat            89985 non-null  float64
 6   start_lon            89985 non-null  float64
 7   end_station          92124 non-null  int64  
 8   end_lat              88052 non-null  float64
 9   end_lon              88052 non-null  float64
 10  bike_id              92124 non-null  object 
 11  plan_duration        92124 non-null  int64  
 12  trip_route_category  92124 non-null  object 
 13  passholder_type      92124 non-null  object 
 14  bike_type            92124 non-null  object 
dtypes: float64(4), int64(5), object(6)
m

Null values in start lat, start lon, end lat, end lon

#### Checking count of null values

In [6]:
metro.isnull().sum()

trip_id                   0
duration                  0
start_time                0
end_time                  0
start_station             0
start_lat              2139
start_lon              2139
end_station               0
end_lat                4072
end_lon                4072
bike_id                   0
plan_duration             0
trip_route_category       0
passholder_type           0
bike_type                 0
dtype: int64

To get rid of null values there are 2 options:
1. Remove rows (min 4072 or above rows can be deleted)
2. Drop lat and lon coloumns (4 cols)

### Removing rows with null values

In [7]:
mt_copy= metro

In [8]:
mt_copy = mt_copy.dropna(how='any',axis=0)

mt_copy.info()

In [9]:
mt_copy.duplicated().any()

False

## No Duplicates as the function returned false

In [10]:
mt_copy.describe()

trip_id      duration  start_station     start_lat     start_lon  \
count  8.676000e+04  86760.000000   86760.000000  86760.000000  86760.000000   
mean   1.274167e+08     26.996404    3445.616367     34.035340   -118.283984   
std    1.526039e+06     95.720579     597.544205      0.058808      0.069444   
min    1.246571e+08      1.000000    3005.000000     33.710979   -118.495422   
25%    1.261132e+08      6.000000    3028.000000     34.038609   -118.279762   
50%    1.274679e+08     11.000000    3052.000000     34.046822   -118.258537   
75%    1.287317e+08     21.000000    4250.000000     34.051941   -118.248253   
max    1.303877e+08   1440.000000    4453.000000     34.177662   -118.231277   

        end_station       end_lat       end_lon  plan_duration  
count  86760.000000  86760.000000  86760.000000   86760.000000  
mean    3449.219006     34.035114   -118.284264      59.330440  
std      598.836056      0.058812      0.070127     108.474264  
min     3005.000000     33.710979   -118.495422       1.000000  
25%     3028.000000     34.038609   -118.280182      30.000000  
50%     3052.000000     34.046810   -118.258537      30.000000  
75%     4250.000000     34.051941   -118.247490      30.000000  
max     4453.000000     34.177662   -118.231277     999.000000

In [11]:
mt_copy["plan_duration"].describe()

count    86760.000000
mean        59.330440
std        108.474264
min          1.000000
25%         30.000000
50%         30.000000
75%         30.000000
max        999.000000
Name: plan_duration, dtype: float64

In [12]:
mt_copy["plan_duration"].value_counts()

30     55907
1      21451
365     9375
999       27
Name: plan_duration, dtype: int64

In [13]:
mt_copy.groupby("passholder_type")["plan_duration"].describe()

count        mean        std    min    25%    50%    75%  \
passholder_type                                                               
Annual Pass       5966.0  365.000000    0.00000  365.0  365.0  365.0  365.0   
Flex Pass            6.0  365.000000    0.00000  365.0  365.0  365.0  365.0   
Monthly Pass     55904.0   30.000000    0.00000   30.0   30.0   30.0   30.0   
One Day Pass      3599.0    1.000000    0.00000    1.0    1.0    1.0    1.0   
Testing             27.0  999.000000    0.00000  999.0  999.0  999.0  999.0   
Walk-up          21258.0   59.273544  133.47371    1.0    1.0    1.0    1.0   

                   max  
passholder_type         
Annual Pass      365.0  
Flex Pass        365.0  
Monthly Pass      30.0  
One Day Pass       1.0  
Testing          999.0  
Walk-up          365.0

In [14]:
mt_copy.groupby("passholder_type")["trip_route_category"].describe()

count unique         top   freq
passholder_type                                 
Annual Pass       5966      2     One Way   5621
Flex Pass            6      1     One Way      6
Monthly Pass     55904      2     One Way  49866
One Day Pass      3599      2     One Way   2291
Testing             27      2  Round Trip     21
Walk-up          21258      2     One Way  14239

###### Count: Total number of pass holders acc. to type
###### Unique: No. of unique values in each category ( One way/ Round)
###### Top: Frequent in that category
###### Freq: Frequency of top type (check)

In [15]:
mt_copy.groupby("passholder_type")["bike_type"].describe()

count unique       top   freq
passholder_type                               
Annual Pass       5966      3  electric   3344
Flex Pass            6      2  standard      4
Monthly Pass     55904      3  electric  35311
One Day Pass      3599      3     smart   1667
Testing             27      1     smart     27
Walk-up          21258      3  standard   8830

In [16]:
mt_copy["bike_type"].unique()

array(['standard', 'electric', 'smart'], dtype=object)

In [17]:
mt_copy.groupby("bike_type")["passholder_type"].describe()

count unique           top   freq
bike_type                                   
electric   45818      5  Monthly Pass  35311
smart      11976      5       Walk-up   6286
standard   28966      5  Monthly Pass  16798

###### Type of pass held by different bike types
###### Top pass category for electric & standard is monthly pass

 So the first step is to convert them to date and time format using the function to_datetime() from the Pandas library. (can do)

#### Checking abnormality in trip duration

In [18]:
outliers_trip_duration = (mt_copy["duration"] < 1 | (mt_copy["duration"] > 480)) 
outliers_trip_duration.value_counts()

False    86760
Name: duration, dtype: int64

In [19]:
mt_copy.groupby("passholder_type")["duration"].max()

passholder_type
Annual Pass     1440
Flex Pass          9
Monthly Pass    1440
One Day Pass    1440
Testing          292
Walk-up         1440
Name: duration, dtype: int64

#### Consider a particular duration as a threshold and check for outliners

Checked for trips with duration > 12 hrs (720 mins) to find outliners

In [20]:
mt_copy[mt_copy["duration"] >= 720].count()

trip_id                456
duration               456
start_time             456
end_time               456
start_station          456
start_lat              456
start_lon              456
end_station            456
end_lat                456
end_lon                456
bike_id                456
plan_duration          456
trip_route_category    456
passholder_type        456
bike_type              456
dtype: int64

### Calculating the distance travelled by users

In [21]:
mt_copy["trip_distance"] = np.sqrt(((mt_copy["end_lat"] - mt_copy["start_lat"])**2) + ((mt_copy["end_lon"] - mt_copy["start_lon"])**2))


C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3146619345.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["trip_distance"] = np.sqrt(((mt_copy["end_lat"] - mt_copy["start_lat"])**2) + ((mt_copy["end_lon"] - mt_copy["start_lon"])**2))
C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3146619345.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["trip_distance"] = mt_copy["trip_distance"] * 111.139


In [ ]:
mt_copy["trip_distance"] = mt_copy["trip_distance"] * 111.139

In [22]:
mt_copy.head()

trip_id  duration     start_time       end_time  start_station  \
0  124657107         5  7/1/2019 0:04  7/1/2019 0:09           4312   
1  124657587         9  7/1/2019 0:07  7/1/2019 0:16           3066   
2  124658068         5  7/1/2019 0:20  7/1/2019 0:25           4410   
3  124659747        20  7/1/2019 0:44  7/1/2019 1:04           3045   
4  124660227        27  7/1/2019 0:44  7/1/2019 1:11           3035   

   start_lat   start_lon  end_station    end_lat     end_lon bike_id  \
0  34.066990 -118.290878         4410  34.063351 -118.296799    6168   
1  34.063389 -118.236160         3066  34.063389 -118.236160   17584   
2  34.063351 -118.296799         4312  34.066990 -118.290878   18920   
3  34.028511 -118.256668         4275  34.012520 -118.285896    6016   
4  34.048401 -118.260948         3049  34.056969 -118.253593    5867   

   plan_duration trip_route_category passholder_type bike_type  trip_distance  
0             30             One Way    Monthly Pass  standard       0.772401  
1             30          Round Trip    Monthly Pass  electric       0.000000  
2             30             One Way    Monthly Pass  electric       0.772401  
3              1             One Way         Walk-up  standard       3.702761  
4             30             One Way    Monthly Pass  standard       1.254969

In [23]:
mt_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 86760 entries, 0 to 92122
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   trip_id              86760 non-null  int64  
 1   duration             86760 non-null  int64  
 2   start_time           86760 non-null  object 
 3   end_time             86760 non-null  object 
 4   start_station        86760 non-null  int64  
 5   start_lat            86760 non-null  float64
 6   start_lon            86760 non-null  float64
 7   end_station          86760 non-null  int64  
 8   end_lat              86760 non-null  float64
 9   end_lon              86760 non-null  float64
 10  bike_id              86760 non-null  object 
 11  plan_duration        86760 non-null  int64  
 12  trip_route_category  86760 non-null  object 
 13  passholder_type      86760 non-null  object 
 14  bike_type            86760 non-null  object 
 15  trip_distance        86760 non-null 

In [24]:
mt_copy["start_time"] = pd.to_datetime(mt_copy["start_time"])
mt_copy["end_time"] = pd.to_datetime(mt_copy["end_time"])

C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3072117926.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["start_time"] = pd.to_datetime(mt_copy["start_time"])
C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3072117926.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["end_time"] = pd.to_datetime(mt_copy["end_time"])


In [25]:
mt_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 86760 entries, 0 to 92122
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   trip_id              86760 non-null  int64         
 1   duration             86760 non-null  int64         
 2   start_time           86760 non-null  datetime64[ns]
 3   end_time             86760 non-null  datetime64[ns]
 4   start_station        86760 non-null  int64         
 5   start_lat            86760 non-null  float64       
 6   start_lon            86760 non-null  float64       
 7   end_station          86760 non-null  int64         
 8   end_lat              86760 non-null  float64       
 9   end_lon              86760 non-null  float64       
 10  bike_id              86760 non-null  object        
 11  plan_duration        86760 non-null  int64         
 12  trip_route_category  86760 non-null  object        
 13  passholder_type      86760 non-

In [26]:
mt_copy["start_hour"] = mt_copy["start_time"].dt.hour
mt_copy["start_hour"].unique()

C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\1171425692.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["start_hour"] = mt_copy["start_time"].dt.hour


array([ 0,  1,  2,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23,  3], dtype=int64)

In [ ]:
mt_copy["start_day_num"] = mt_copy["start_time"].dt.dayofweek
day_dict = {0:"Sunday", 1:"Monday", 2:"Tuesday", 3:"Wednesday", 4:"Thursday", 5:"Friday", 6:"Saturday"}
mt_copy["start_day_name"]= mt_copy["start_day_num"].apply(lambda y:day_dict[y])

#### Change the day_dict (numbering of days) and change the condition in later code!!!

In [28]:
mt_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 86760 entries, 0 to 92122
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   trip_id              86760 non-null  int64         
 1   duration             86760 non-null  int64         
 2   start_time           86760 non-null  datetime64[ns]
 3   end_time             86760 non-null  datetime64[ns]
 4   start_station        86760 non-null  int64         
 5   start_lat            86760 non-null  float64       
 6   start_lon            86760 non-null  float64       
 7   end_station          86760 non-null  int64         
 8   end_lat              86760 non-null  float64       
 9   end_lon              86760 non-null  float64       
 10  bike_id              86760 non-null  object        
 11  plan_duration        86760 non-null  int64         
 12  trip_route_category  86760 non-null  object        
 13  passholder_type      86760 non-

In [29]:
mt_copy.head(10)

trip_id  duration          start_time            end_time  start_station  \
0  124657107         5 2019-07-01 00:04:00 2019-07-01 00:09:00           4312   
1  124657587         9 2019-07-01 00:07:00 2019-07-01 00:16:00           3066   
2  124658068         5 2019-07-01 00:20:00 2019-07-01 00:25:00           4410   
3  124659747        20 2019-07-01 00:44:00 2019-07-01 01:04:00           3045   
4  124660227        27 2019-07-01 00:44:00 2019-07-01 01:11:00           3035   
5  124659507         7 2019-07-01 00:51:00 2019-07-01 00:58:00           3056   
6  124662392        37 2019-07-01 01:22:00 2019-07-01 01:59:00           3047   
7  124663117         8 2019-07-01 02:06:00 2019-07-01 02:14:00           3069   
8  124664810         4 2019-07-01 02:43:00 2019-07-01 02:47:00           3014   
9  124669422         4 2019-07-01 04:17:00 2019-07-01 04:21:00           3075   

   start_lat   start_lon  end_station    end_lat     end_lon bike_id  \
0  34.066990 -118.290878         4410  34.063351 -118.296799    6168   
1  34.063389 -118.236160         3066  34.063389 -118.236160   17584   
2  34.063351 -118.296799         4312  34.066990 -118.290878   18920   
3  34.028511 -118.256668         4275  34.012520 -118.285896    6016   
4  34.048401 -118.260948         3049  34.056969 -118.253593    5867   
5  34.037460 -118.265381         3031  34.044701 -118.252441   18530   
6  34.039982 -118.266403         3047  34.039982 -118.266403   18877   
7  34.050880 -118.248253         3022  34.046070 -118.233093   18984   
8  34.056610 -118.237213         3030  34.051941 -118.243530   12126   
9  34.042110 -118.256187         3064  34.046810 -118.256981   12398   

   plan_duration trip_route_category passholder_type bike_type  trip_distance  \
0             30             One Way    Monthly Pass  standard       0.772401   
1             30          Round Trip    Monthly Pass  electric       0.000000   
2             30             One Way    Monthly Pass  electric       0.772401   
3              1             One Way         Walk-up  standard       3.702761   
4             30             One Way    Monthly Pass  standard       1.254969   
5             30             One Way    Monthly Pass  electric       1.647992   
6             30          Round Trip    Monthly Pass  electric       0.000000   
7             30             One Way    Monthly Pass  electric       1.767640   
8             30             One Way    Monthly Pass  standard       0.873018   
9             30             One Way    Monthly Pass  standard       0.529755   

   start_hour  start_day_num start_day_name  
0           0              0         Sunday  
1           0              0         Sunday  
2           0              0         Sunday  
3           0              0         Sunday  
4           0              0         Sunday  
5           0              0         Sunday  
6           1              0         Sunday  
7           2              0         Sunday  
8           2              0         Sunday  
9           4              0         Sunday

In [31]:
mt_copy["start_day_name"].unique()

array(['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday',
       'Saturday'], dtype=object)

In [42]:
mt_copy["weekend"] = np.where(((mt_copy["start_day_num"] == 6) | (mt_copy["start_day_num"] == 0)) , True, False)

C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3960194967.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["weekend"] = np.where(((mt_copy["start_day_num"] == 6) | (mt_copy["start_day_num"] == 0)) , True, False)


In [43]:
mt_copy.head(60)

trip_id  duration          start_time            end_time  \
0   124657107         5 2019-07-01 00:04:00 2019-07-01 00:09:00   
1   124657587         9 2019-07-01 00:07:00 2019-07-01 00:16:00   
2   124658068         5 2019-07-01 00:20:00 2019-07-01 00:25:00   
3   124659747        20 2019-07-01 00:44:00 2019-07-01 01:04:00   
4   124660227        27 2019-07-01 00:44:00 2019-07-01 01:11:00   
5   124659507         7 2019-07-01 00:51:00 2019-07-01 00:58:00   
6   124662392        37 2019-07-01 01:22:00 2019-07-01 01:59:00   
7   124663117         8 2019-07-01 02:06:00 2019-07-01 02:14:00   
8   124664810         4 2019-07-01 02:43:00 2019-07-01 02:47:00   
9   124669422         4 2019-07-01 04:17:00 2019-07-01 04:21:00   
10  124669421         4 2019-07-01 04:18:00 2019-07-01 04:22:00   
12  124672806        17 2019-07-01 05:16:00 2019-07-01 05:33:00   
13  124673775        18 2019-07-01 05:34:00 2019-07-01 05:52:00   
14  124678322        98 2019-07-01 05:47:00 2019-07-01 07:25:00   
15  124673773         5 2019-07-01 05:48:00 2019-07-01 05:53:00   
16  124674015         6 2019-07-01 05:50:00 2019-07-01 05:56:00   
17  124674986        24 2019-07-01 05:51:00 2019-07-01 06:15:00   
18  124674257         6 2019-07-01 05:55:00 2019-07-01 06:01:00   
19  124674256         6 2019-07-01 05:58:00 2019-07-01 06:04:00   
20  124674982        11 2019-07-01 06:06:00 2019-07-01 06:17:00   
21  124674981         6 2019-07-01 06:12:00 2019-07-01 06:18:00   
22  124675466        16 2019-07-01 06:13:00 2019-07-01 06:29:00   
23  124675222         6 2019-07-01 06:17:00 2019-07-01 06:23:00   
24  124675223         7 2019-07-01 06:17:00 2019-07-01 06:24:00   
25  124675221         4 2019-07-01 06:18:00 2019-07-01 06:22:00   
26  124675945        16 2019-07-01 06:22:00 2019-07-01 06:38:00   
27  124675704         3 2019-07-01 06:28:00 2019-07-01 06:31:00   
28  124675943         9 2019-07-01 06:29:00 2019-07-01 06:38:00   
29  124675942         7 2019-07-01 06:30:00 2019-07-01 06:37:00   
30  124676184        10 2019-07-01 06:30:00 2019-07-01 06:40:00   
31  124676652        21 2019-07-01 06:30:00 2019-07-01 06:51:00   
32  124676182         9 2019-07-01 06:31:00 2019-07-01 06:40:00   
33  124675941         2 2019-07-01 06:35:00 2019-07-01 06:37:00   
34  124676180         5 2019-07-01 06:38:00 2019-07-01 06:43:00   
35  124676648        13 2019-07-01 06:39:00 2019-07-01 06:52:00   
36  124678066        44 2019-07-01 06:40:00 2019-07-01 07:24:00   
37  124676414         6 2019-07-01 06:41:00 2019-07-01 06:47:00   
38  124677112        19 2019-07-01 06:41:00 2019-07-01 07:00:00   
39  124677834        34 2019-07-01 06:41:00 2019-07-01 07:15:00   
41  124676413         3 2019-07-01 06:44:00 2019-07-01 06:47:00   
42  124676645         5 2019-07-01 06:47:00 2019-07-01 06:52:00   
43  124676644         4 2019-07-01 06:48:00 2019-07-01 06:52:00   
44  124676643         5 2019-07-01 06:49:00 2019-07-01 06:54:00   
45  124676874         6 2019-07-01 06:50:00 2019-07-01 06:56:00   
46  124677342        16 2019-07-01 06:50:00 2019-07-01 07:06:00   
47  124676642         3 2019-07-01 06:51:00 2019-07-01 06:54:00   
48  124677339        14 2019-07-01 06:53:00 2019-07-01 07:07:00   
49  124677338        10 2019-07-01 06:55:00 2019-07-01 07:05:00   
50  124677104         5 2019-07-01 06:58:00 2019-07-01 07:03:00   
51  124678045        19 2019-07-01 07:01:00 2019-07-01 07:20:00   
53  124678044        19 2019-07-01 07:03:00 2019-07-01 07:22:00   
54  124677574         6 2019-07-01 07:06:00 2019-07-01 07:12:00   
55  124677811        13 2019-07-01 07:06:00 2019-07-01 07:19:00   
56  124677571         5 2019-07-01 07:07:00 2019-07-01 07:12:00   
57  124677570         3 2019-07-01 07:09:00 2019-07-01 07:12:00   
58  124677807         4 2019-07-01 07:13:00 2019-07-01 07:17:00   
59  124678035        10 2019-07-01 07:13:00 2019-07-01 07:23:00   
60  124678734        25 2019-07-01 07:14:00 2019-07-01 07:39:00   
61  124677806         3 2019-07-01 07:15:00 2019-07-0

In [44]:
mt_copy.loc[5000,:]

trip_id                          124986041
duration                                 1
start_time             2019-07-06 12:54:00
end_time               2019-07-06 12:55:00
start_station                         3042
start_lat                        34.049301
start_lon                      -118.238808
end_station                           3042
end_lat                          34.049301
end_lon                        -118.238808
bike_id                              17506
plan_duration                           30
trip_route_category             Round Trip
passholder_type               Monthly Pass
bike_type                         electric
trip_distance                          0.0
start_hour                              12
start_day_num                            5
start_day_name                      Friday
weekend                              False
Name: 5000, dtype: object

In [48]:
mt_copy[mt_copy["start_day_name"] == "Wednesday"].head()

trip_id  duration          start_time            end_time  \
2918  124843599         3 2019-07-04 00:06:00 2019-07-04 00:09:00   
2919  124892784       973 2019-07-04 00:28:00 2019-07-04 16:41:00   
2920  124843598         8 2019-07-04 00:33:00 2019-07-04 00:41:00   
2921  124843855       105 2019-07-04 00:35:00 2019-07-04 02:20:00   
2922  124843854        21 2019-07-04 00:44:00 2019-07-04 01:05:00   

      start_station  start_lat   start_lon  end_station    end_lat  \
2918           3008  34.046612 -118.262733         3064  34.046810   
2919           4405  34.051899 -118.293678         4405  34.051899   
2920           3038  34.046822 -118.248352         3075  34.042110   
2921           4410  34.063351 -118.296799         4410  34.063351   
2922           4227  34.029800 -118.270309         3063  34.048038   

         end_lon bike_id  plan_duration trip_route_category passholder_type  \
2918 -118.256981   18530             30             One Way    Monthly Pass   
2919 -118.293678   16385            365          Round Trip         Walk-up   
2920 -118.256187    5913            365             One Way     Annual Pass   
2921 -118.296799   17493             30          Round Trip    Monthly Pass   
2922 -118.253738   12358              1             One Way         Walk-up   

     bike_type  trip_distance  start_hour  start_day_num start_day_name  \
2918  electric       0.639650           0              3      Wednesday   
2919  electric       0.000000           0              3      Wednesday   
2920  standard       1.016118           0              3      Wednesday   
2921  electric       0.000000           0              3      Wednesday   
2922  standard       2.738675           0              3      Wednesday   

      weekend  
2918    False  
2919    False  
2920    False  
2921    False  
2922    False

In [49]:
mt_copy['start_month_num'] = mt_copy['start_time'].dt.month
mt_copy["start_month_num"].unique()

C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\2252108594.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy['start_month_num'] = mt_copy['start_time'].dt.month


array([7, 8, 9], dtype=int64)

In [50]:
mt_copy.head(-10)

trip_id  duration          start_time            end_time  \
0      124657107         5 2019-07-01 00:04:00 2019-07-01 00:09:00   
1      124657587         9 2019-07-01 00:07:00 2019-07-01 00:16:00   
2      124658068         5 2019-07-01 00:20:00 2019-07-01 00:25:00   
3      124659747        20 2019-07-01 00:44:00 2019-07-01 01:04:00   
4      124660227        27 2019-07-01 00:44:00 2019-07-01 01:11:00   
...          ...       ...                 ...                 ...   
92108  130053104        13 2019-09-30 22:42:00 2019-09-30 22:55:00   
92109  130053102         6 2019-09-30 22:43:00 2019-09-30 22:49:00   
92110  130053101         5 2019-09-30 22:56:00 2019-09-30 23:01:00   
92111  130053100       151 2019-09-30 22:59:00 2019-10-01 01:30:00   
92112  130161467      1440 2019-09-30 22:59:00 2019-10-02 06:17:00   

       start_station  start_lat   start_lon  end_station    end_lat  \
0               4312  34.066990 -118.290878         4410  34.063351   
1               3066  34.063389 -118.236160         3066  34.063389   
2               4410  34.063351 -118.296799         4312  34.066990   
3               3045  34.028511 -118.256668         4275  34.012520   
4               3035  34.048401 -118.260948         3049  34.056969   
...              ...        ...         ...          ...        ...   
92108           4432  34.139568 -118.362251         4426  34.149872   
92109           4390  34.069271 -118.296593         4304  34.062580   
92110           4420  34.168629 -118.377068         4413  34.158321   
92111           4345  33.995560 -118.481552         4345  33.995560   
92112           3011  34.041130 -118.267982         3005  34.048500   

          end_lon  ... plan_duration  trip_route_category passholder_type  \
0     -118.296799  ...            30              One Way    Monthly Pass   
1     -118.236160  ...            30           Round Trip    Monthly Pass   
2     -118.290878  ...            30              One Way    Monthly Pass   
3     -118.285896  ...             1              One Way         Walk-up   
4     -118.253593  ...            30              One Way    Monthly Pass   
...           ...  ...           ...                  ...             ...   
92108 -118.378891  ...            30              One Way    Monthly Pass   
92109 -118.290092  ...            30              One Way    Monthly Pass   
92110 -118.370811  ...           365              One Way         Walk-up   
92111 -118.481552  ...             1           Round Trip         Walk-up   
92112 -118.258537  ...             1              One Way    One Day Pass   

      bike_type trip_distance  start_hour  start_day_num  start_day_name  \
0      standard      0.772401           0              0          Sunday   
1      electric      0.000000           0              0          Sunday   
2      electric      0.772401           0              0          Sunday   
3      standard      3.702761           0              0          Sunday   
4      standard      1.254969           0              0          Sunday   
...         ...           ...         ...            ...             ...   
92108     smart      2.175209          22              0          Sunday   
92109  standard      1.036829          22              0          Sunday   
92110     smart      1.340158          22              0          Sunday   
92111     smart      0.000000          22              0          Sunday   
92112  electric      1.331466          22              0          Sunday   

      weekend  start_month_num  
0        True                7  
1        True                7  
2        True                7  
3        True                7  
4        True                7  
...       ...              ...  
92108    True                9  
92109    True                9  
92110    True                9  
92111    True                9  
92112    True                9  

[86750 rows x 21 columns]

In [51]:
metro.head(10)

trip_id  duration     start_time       end_time  start_station  \
0  124657107         5  7/1/2019 0:04  7/1/2019 0:09           4312   
1  124657587         9  7/1/2019 0:07  7/1/2019 0:16           3066   
2  124658068         5  7/1/2019 0:20  7/1/2019 0:25           4410   
3  124659747        20  7/1/2019 0:44  7/1/2019 1:04           3045   
4  124660227        27  7/1/2019 0:44  7/1/2019 1:11           3035   
5  124659507         7  7/1/2019 0:51  7/1/2019 0:58           3056   
6  124662392        37  7/1/2019 1:22  7/1/2019 1:59           3047   
7  124663117         8  7/1/2019 2:06  7/1/2019 2:14           3069   
8  124664810         4  7/1/2019 2:43  7/1/2019 2:47           3014   
9  124669422         4  7/1/2019 4:17  7/1/2019 4:21           3075   

   start_lat   start_lon  end_station    end_lat     end_lon bike_id  \
0  34.066990 -118.290878         4410  34.063351 -118.296799    6168   
1  34.063389 -118.236160         3066  34.063389 -118.236160   17584   
2  34.063351 -118.296799         4312  34.066990 -118.290878   18920   
3  34.028511 -118.256668         4275  34.012520 -118.285896    6016   
4  34.048401 -118.260948         3049  34.056969 -118.253593    5867   
5  34.037460 -118.265381         3031  34.044701 -118.252441   18530   
6  34.039982 -118.266403         3047  34.039982 -118.266403   18877   
7  34.050880 -118.248253         3022  34.046070 -118.233093   18984   
8  34.056610 -118.237213         3030  34.051941 -118.243530   12126   
9  34.042110 -118.256187         3064  34.046810 -118.256981   12398   

   plan_duration trip_route_category passholder_type bike_type  
0             30             One Way    Monthly Pass  standard  
1             30          Round Trip    Monthly Pass  electric  
2             30             One Way    Monthly Pass  electric  
3              1             One Way         Walk-up  standard  
4             30             One Way    Monthly Pass  standard  
5             30             One Way    Monthly Pass  electric  
6             30          Round Trip    Monthly Pass  electric  
7             30             One Way    Monthly Pass  electric  
8             30             One Way    Monthly Pass  standard  
9             30             One Way    Monthly Pass  standard

In [56]:
metro["start_time"][0]

'7/1/2019 0:04'

In [57]:
month_dict = {1:"January", 2:"February", 3:"March", 4:"April", 5:"May", 6:"June", 7:"July", 8:"August", 9:"September", 10:"October", 11:"November", 12:"December"}
mt_copy["start_month_name"]= mt_copy["start_month_num"].apply(lambda y:month_dict[y])
mt_copy["start_month_name"].unique()

C:\Users\shrut\AppData\Local\Temp\ipykernel_13788\3227155754.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  mt_copy["start_month_name"]= mt_copy["start_month_num"].apply(lambda y:month_dict[y])


array(['July', 'August', 'September'], dtype=object)